examples_tests/61.BlitFilterTest/main.cpp

```diff
@@ -0,0 +1,1561 @@
+	{
+		outPC.outDim = { outImageExtent.x, outImageExtent.y, outImageExtent.z };
+		outPC.inPixelCount = inImageExtent.x * inImageExtent.y * inImageExtent.z;
+		outPC.referenceAlpha = static_cast<float>(referenceAlpha);
+
+		const core::vectorSIMDu32 workgroupSize(NBL_GLSL_DEFAULT_WORKGROUP_DIM, NBL_GLSL_DEFAULT_WORKGROUP_DIM, NBL_GLSL_DEFAULT_WORKGROUP_DIM, 1u);
+		const core::vectorSIMDu32 workgroupCount = (outImageExtent + workgroupSize - core::vectorSIMDu32(1u, 1u, 1u, 1u)) / workgroupSize;
+
+		outDispatchInfo.wgCount[0] = workgroupCount.x;
+		outDispatchInfo.wgCount[1] = workgroupCount.y;
+		outDispatchInfo.wgCount[2] = workgroupCount.z;
+	}
+
+	core::smart_refctd_ptr<video::IGPUSpecializedShader> createBlitSpecializedShader(const asset::E_FORMAT inFormat, const asset::E_FORMAT outFormat, const asset::IImage::E_TYPE inImageType,
+		const core::vectorSIMDu32& inExtent, const core::vectorSIMDu32& outExtent, const BlitFilter::CState::E_ALPHA_SEMANTIC alphaSemantic)
+	{
+		const char* shaderpath = "../default_compute_blit.comp";
+		auto cpuShader = loadShaderFromFile(shaderpath);
+		if (!cpuShader)
+			return nullptr;
+
+		const uint32_t inChannelCount = asset::getFormatChannelCount(inFormat);
+		const uint32_t outChannelCount = asset::getFormatChannelCount(outFormat);
+		assert(outChannelCount <= inChannelCount);
+
+		core::vectorSIMDf scale = static_cast<core::vectorSIMDf>(inExtent).preciseDivision(static_cast<core::vectorSIMDf>(outExtent));
+
+		const core::vectorSIMDu32 windowDim = static_cast<core::vectorSIMDu32>(core::ceil(scale));
+		const uint32_t windowPixelCount = windowDim.x * windowDim.y * windowDim.z;
+		// It is important not to use asset::getTexelOrBlockBytesize here.
+		// Even though input pixels are stored in the shared memory we use outChannelCount here because only outChannelCount channels of
+		// the input image actually need blitting and hence only those channels are stored in shared memory.
+		const size_t windowSize = static_cast<size_t>(windowPixelCount) * outChannelCount * sizeof(float);
+		// Fail if the window cannot be preloaded into shared memory
+		if (windowSize > sharedMemorySize)
+			return nullptr;
+
+		// Fail if I would need to reuse invocations just to process a single window
+		if (windowPixelCount > NBL_GLSL_DEFAULT_WORKGROUP_SIZE)
+			return nullptr;
+
+		// inFormat should support SAMPLED_BIT format feature
+
+		const asset::E_FORMAT outImageViewFormat = getOutImageViewFormat(outFormat);
+		if (outImageViewFormat == asset::EF_UNKNOWN)
+			return nullptr;
+
+		const char* outImageFormatGLSLString = getGLSLFormatStringFromFormat(outFormat);
+		const char* outImageViewFormatGLSLString = getGLSLFormatStringFromFormat(outImageViewFormat);
+
+		char formatInclude[1024] = "";
+		if (outFormat != outImageViewFormat)
+			snprintf(formatInclude, sizeof(formatInclude), "#include <../blit/formats/%s.glsl>\n", outImageFormatGLSLString);
```

Comment:
I don't really like the `outImageFormatGLSLString` shenanigans, because your switch only gives out names of `image2D` layout format qualifiers

but your software encode should handle formats for which no such qualifier exists, such as RGBA8_SRGB, etc.

Reply:
How will you create a storage image with RGBA8_SRGB?

Reply:
Okay you can create a storage image with `RGBA8_SRGB` format as long as it has the `EXTENDED_USAGE_BIT` (in the create flags) set.

I also now understand what the problem is here.

Maybe we can just think of the return value of `getGLSLFormatStringFromFormat` as filenames because they are really not used as layout format qualifiers anywhere (there is `outImageViewFormatGLSLString` for that). So for formats like RGBA8_SRGB we could just return `rgba8_srgb`?

Reply:
I'd return the format that user needs to create the image view with so he can compare the formats (original vs needed one)

Reply:
Then how will you pick the encode file when creating the shader?

Reply:
instead of having different encode files (its getting a bit unwieldy), I'd rather have one file which picks the function impl based on injected preprocessor directive value

Reply:
Resolved in the following commit.

Reply:
Pushed.